### PR TITLE
[#8] [Backend] As a user, I can register an account

### PR DESCRIPTION
**Pull request by mosharaf13**
Closes https://github.com/mosharaf13/google-scrapper-ruby/issues/8
Closes https://github.com/mosharaf13/google-scrapper-ruby/issues/5
Closes https://github.com/mosharaf13/google-scrapper-ruby/issues/4 

## What happened 👀

- Add backend for user registration system.
- Show error message if password doesn't satisfy password requirement (e.g; min 8 chars, uppercase + lowercase + number)
- Redirect to homepage after successful signup

## Insight 📝

- Added devise gem for user authentication.
- Added regex validation to check password complexity

## Proof Of Work 📹

<details>
  <summary>
      - Minimum requirement of 8 chars with mix of uppercase, lowercase, and number
  </summary>

Screen Shot 2023-05-19 at 2 55 05 PM

</details>

<details>
  <summary>
      - Password confirmation must match password field
  </summary>

Screen Shot 2023-05-19 at 2 55 26 PM

</details>

<details>
  <summary>
      - Successful sign up 
  </summary>

Screen Shot 2023-05-19 at 2 56 30 PM

</details>

<details>
  <summary>
      - Successful logout
  </summary>

Screen Shot 2023-05-23 at 9 35 50 AM

Screen Shot 2023-05-23 at 9 36 04 AM

</details>

<details>
  <summary>
      - Successful login
  </summary>
Screen Shot 2023-05-23 at 9 42 04 AM

Screen Shot 2023-05-23 at 9 42 16 AM

</details>

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -15,6 +15,7 @@ gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby] # Windows doe
 # gem 'redis' # Use Redis adapter to run Action Cable in production
 # gem 'kredis' # Use Kredis to get higher-level data types in Redis
 # gem 'bcrypt' # Use Active Model has_secure_password
+gem 'devise'
```
Comment (longnd) [inline Gemfile:18]: let's be consistent with other gem
```suggestion
gem 'devise' # Flexible authentication solution for Rails with Warden
```

File: db/migrate/20230519045115_devise_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+class DeviseCreateUsers < ActiveRecord::Migration[7.0]
+  def change
+    create_table :users do |t|
+      ## Database authenticatable
+      t.string :email,              null: false, default: ""
```
Comment (longnd) [inline db/migrate/20230519045115_devise_create_users.rb:7]: it should use citext data type
https://nimblehq.co/compass/development/web/database/#data-types

**Review by longnd — COMMENTED** [review]
Just a quick first review, I will complete the review this afternoon 🙏
Could you please add a Github .CODEOWNERS file to include Sang, Top, you & I so each of us will receive notification when the PR is ready for review?
https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners

example:
https://github.com/nimblehq/rails-templates/blob/develop/.github/CODEOWNERS

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
```
Comment (sanG-github) [inline app/models/user.rb:6]: As this model will be bigger and bigger as time flies. So, we usually extract these configurations into a Module, and then we can include them back into this model.

```ruby
# models/user.rb
class User < ApplicationRecord
  include Authenticable
end

# models/concerns/authenticable
module Authenticable
  extend ActiveSupport::Concern

  included do
    # Include default devise modules. Others available are:
    # :confirmable, :lockable, :timeoutable and :omniauthable
    devise :database_authenticatable, :registerable, :recoverable, :rememberable, :validatable

    validate :password_complexity, if: :password_required?
  end
end
```

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  # Other Devise configurations
+
+  validate :password_complexity, if: :password_required?
+
+  def password_complexity
+    return if password.blank? || password.match?(/(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}/)
+
+    errors.add(:password, 'must include at least one uppercase letter, one lowercase letter, and one digit')
+  end
```
Comment (sanG-github) [inline app/models/user.rb:17]: As we've already checked it with `password_required?`, do you think we need this check? 
Also, we can define the PASSWORD_PATTERN as a constant.
Moreover, we can add a custom `type` for the `error`, and then we can [add the locale for this error](https://guides.rubyonrails.org/v3.2.0/i18n.html#:~:text=For%20example%2C%20you%20might%20have%20an%20Admin%20model%20inheriting%20from%20User%3A) later.

```suggestion
  PASSWORD_PATTERN = /(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}/
  
  def password_complexity
    return if password.match?(PASSWORD_PATTERN)

    errors.add(:password, :invalid)
  end
```

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  private
+
+  def password_required?
+    !persisted? || password.present? || password_confirmation.present?
```
Comment (sanG-github) [inline app/models/user.rb:22]: Did you mean we should check it when creating the user, right? 🙊 
If yes, we can do it with the `on: :create` [option](https://guides.rubyonrails.org/active_record_callbacks.html#:~:text=Callbacks%20can%20also%20be%20registered%20to%20only%20fire%20on%20certain%20life%20cycle%20events%3A). ✨

File: app/views/home/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<% if user_signed_in? %> 
+ <div> Welcome <%= current_user.email %> </div> 
+  <%= button_to "Sign out", destroy_user_session_path, method: :delete %> 
+<% else %>
+  <%= button_to "Sign in", new_user_session_path %> 
+<% end %>
+<h1>
+    Welcome to the world of scrapping!!
```
Comment (sanG-github) [inline app/views/home/index.html.erb:8]: Make sure we have the `same indent`, and always have the EOF line. 🚀

**Review by sanG-github — CHANGES_REQUESTED** [review]
Some first suggestions 🚀

File: db/migrate/20230519045115_devise_create_users.rb (hunk 2 above)
Comment (mosharaf13) [reply db/migrate/20230519045115_devise_create_users.rb:7]: Hi @longnd. Thanks for the suggestion 🙏🏼. I have changed the type of email field.

File: app/models/user.rb (hunk 3 above)
Comment (mosharaf13) [reply app/models/user.rb]: Do we usually apply this type of module based pattern to other classes as well? e.g. controllers, services. If so, I will keep that in mind next time. Also, if you see any pattern that can be optimized, please feel free to  throw some suggestions 😀 .

File: app/models/user.rb (hunk 4 above)
Comment (mosharaf13) [reply app/models/user.rb]: Thanks for this great suggestion 👍🏼 . I tried to add this extra checking for password just to see how it's done in rails philosophy. More dirty hands, more learning 😄. Now going with your suggested `:invalid` message. I'll try to add custom validation message later.

File: app/models/user.rb (hunk 5 above)
Comment (mosharaf13) [reply app/models/user.rb]: Corrected. I have removed password required actually (look below). Thanks for the suggestion 👍🏼 . Is there any large/standard model file you can suggest me to take a look at from nimble projects?

```
class User < ApplicationRecord
  include Authenticable
  PASSWORD_PATTERN = /(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}/

  before_validation :password_complexity, on: :create

  def password_complexity
    return if password.match?(PASSWORD_PATTERN)

    errors.add(:password, :invalid)
  end
end
```

File: app/views/home/index.html.erb (hunk 6 above)
Comment (mosharaf13) [reply app/views/home/index.html.erb]: Added. Thanks 👍🏼 . I need to build habit of running `rubocop -A` before a push 😬 .

**Review by longnd — APPROVED** [review]

File: app/controllers/home_controller.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class HomeController < ApplicationController
+  def index
+    render
+  end
```
Comment (sanG-github) [inline app/controllers/home_controller.rb:6]: We can do it without `render` 🚀 
```suggestion
  def index; end
```

File: app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  include Authenticable
+  PASSWORD_PATTERN = /(?=.*[A-Z])(?=.*[a-z])(?=.*\d).{8,}/
+
+  before_validation :password_complexity, on: :create
+
+  def password_complexity
```
Comment (sanG-github) [inline app/models/user.rb:9]: This method must be private, then, no one can access it out of this validation ✨ 

```suggestion
  private
  
  def password_complexity
```

File: app/views/home/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<% if user_signed_in? %> 
+ <div> Welcome <%= current_user.email %> </div> 
+  <%= button_to "Sign out", destroy_user_session_path, method: :delete %> 
+<% else %>
+  <%= button_to "Sign in", new_user_session_path %> 
+<% end %>
+<h1>
+  Welcome to the world of scrapping!!
+</h1>
```
Comment (sanG-github) [inline app/views/home/index.html.erb:9]: I added:
- 1 space before for the `<div> Welcome <%= current_user.email %> </div>`
- 1 empty line after the `if else statement`
- Make the `h1` element inline
- EOF (end-of-file)

```suggestion
<% if user_signed_in? %> 
  <div> Welcome <%= current_user.email %> </div> 
  <%= button_to "Sign out", destroy_user_session_path, method: :delete %> 
<% else %>
  <%= button_to "Sign in", new_user_session_path %> 
<% end %>

<h1> Welcome to the world of scrapping!! </h1>

```

File: config/routes.rb [hunk 10]
```diff
@@ -1,10 +1,11 @@
...
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
-
+  devise_for :users
+  
   # Defines the root path route ("/")
   # root "articles#index"
-  root "rails/welcome#index"
-
+  root "home#index"
+
```
Comment (sanG-github) [inline config/routes.rb:9]:
```suggestion
  root "home#index"

```

**Review by sanG-github — COMMENTED** [review]
Last round of reviews, we're getting there!!! 🚀

File: app/views/home/index.html.erb (hunk 9 above)
Comment (mosharaf13) [reply app/views/home/index.html.erb:9]: Thanks 😀

**Review by sanG-github — APPROVED** [review]

**Pull request merged**